core/trino-main/src/main/java/io/trino/sql/planner/ExpressionInterpreter.java

```diff
@@ -633,6 +633,14 @@ protected Object visitInPredicate(InPredicate node, Object context)
 
             ResolvedFunction equalsOperator = metadata.resolveOperator(session, OperatorType.EQUAL, types(node.getValue(), valueList));
             for (Expression expression : valueList.getValues()) {
+                if (value instanceof Expression && expression instanceof Literal) {
```

Comment:
> You're skipping the process(expression, context) call.

Yes, `process` is potentially expensive due to parsing of type. 

> This looks like an optimization specifically for bigints (which get encoded as GenericLiteral and may be relatively expensive to process. Maybe we do something like https://github.com/trinodb/trino/pull/11918 instead?

No. It's not targeted at `BIGINTs`. It target `GenericLiteral` cases like
```
IN (VARCHAR 'XX2` VARCHAR 'XX1`)
```
So your fix only target specific case, while mine covers `GenericLiteral` in general.

Reply:
> Yes, `process` is potentially expensive due to parsing of type.

Yes, i also noticed SqlParser on a thraeddump

> No. It's not targeted at `BIGINTs`. It target `GenericLiteral` cases like
> 
> ```
> IN (VARCHAR 'XX2` VARCHAR 'XX1`)
> ```

only an unbounded varchar value will be encoded like this.
A bounded varchar will be encoded with a `CAST`.

>   So your fix only target specific case, while mine covers GenericLiteral in general.

Correct.
Also, this PR addresses only IN list, while the other PR (https://github.com/trinodb/trino/pull/11918) addresses any use of GenericLiteral.

1. we can extend the other PR to cover the types LiteralEncoder uses GenericLiteral for: tinyint, smallint, bigint, varchar, real, date -- only 6 to handle.
2. we can extend GenericLiteral to optionally carry the explicit `Type`
3. we can introduce new literal form(s) with explicit type for the 6 types mentioned above (of subset of them, those which actually matter)

Reply:
@findepi One thing I'm doing with this PR is simple optimization to avoid recursion for simple literal case to unblock users. I don't want to get into details how literals are encoded and parsed. I think optimization is worth it.

> Also, this PR addresses only IN list, while the other PR (https://github.com/trinodb/trino/pull/11918) addresses any use of GenericLiteral.

Yes, but it's long IN lists that are really problematic.

> A bounded varchar will be encoded with a CAST.

It probably should be encoded via `GenericLiteral`, rather than complex expression

> we can extend GenericLiteral to optionally carry the explicit Type

It does carry type as a string.

Reply:
> One thing I'm doing with this PR is simple optimization to avoid recursion for simple literal case to unblock users.

I agree both PRs are tactical improvements, a short-term solution.

How do you benchmark this PR's improvement? I would expect both to be equally efficient.

Reply:
> I agree both PRs are tactical improvements, a short-term solution.

This one doesn't do string comparison compared to your PR.

> How do you benchmark this PR's improvement? I would expect both to be equally efficient.

If these are equally efficient (I don't think they are though), then I don't see a reason of throwing this optimization in favor of https://github.com/trinodb/trino/pull/11918. Unless it's critically bad, but it isn't. So what's the reason?
In fact, both of they might land IMO.